src/gfx/convert.cpp

```diff
@@ -0,0 +1,831 @@
+					bitplanes |= 1;
+				}
+				if (index & 2) {
+					bitplanes |= 0x100;
+				}
+			}
+			_data[y * 2] = bitplanes & 0xFF;
+			_data[y * 2 + 1] = bitplanes >> 8;
+
+			// Update the hash
+			_hash ^= bitplanes;
+			if (options.allowMirroring) {
+				// Count the line itself as mirrorred; vertical mirroring is
+				// already taken care of because the symmetric line will be XOR'd
+				// the same way. (...which is a problem, but probably benign.)
+				_hash ^= flip(bitplanes >> 8) << 8 | flip(bitplanes & 0xFF);
+			}
+		}
+	}
+
+	auto const &data() const { return _data; }
+	uint16_t hash() const { return _hash; }
+
+	enum MatchType {
+		NOPE,
+		EXACT,
+		HFLIP,
+		VFLIP,
+		VHFLIP,
+	};
+	MatchType tryMatching(TileData const &other) const {
+		if (std::equal(_data.begin(), _data.end(), other._data.begin()))
+			return MatchType::EXACT;
+
+		if (options.allowMirroring) {
+			// TODO
+		}
+
+		return MatchType::NOPE;
+	}
+	friend bool operator==(TileData const &lhs, TileData const &rhs) {
+		return lhs.tryMatching(rhs) != MatchType::NOPE;
+	}
+};
+
+template<>
+struct std::hash<TileData> {
+	std::size_t operator()(TileData const &tile) const { return tile.hash(); }
+};
+
+namespace optimized {
+
+struct UniqueTiles {
+	std::unordered_set<TileData> tileset;
+	std::vector<TileData const *> tiles;
+
+	UniqueTiles() = default;
+	// Copies are likely to break pointers, so we really don't want those.
+	// Copy elision should be relied on to be more sure that refs won't be invalidated, too!
+	UniqueTiles(UniqueTiles const &) = delete;
+	UniqueTiles(UniqueTiles &&) = default;
+
+	/**
+	 * Adds a tile to the collection, and returns its ID
+	 */
+	std::tuple<uint16_t, TileData::MatchType> addTile(Png::TilesVisitor::Tile const &tile,
+	                                                  Palette const &palette) {
+		TileData newTile(tile, palette);
+		auto [tileData, inserted] = tileset.insert(newTile);
+
+		TileData::MatchType matchType = TileData::EXACT;
+		if (inserted) {
+			// Give the new tile the next available unique ID
+			tileData->tileID = tiles.size();
+			// Pointers are never invalidated!
+			tiles.emplace_back(&*tileData);
+		} else {
+			matchType = tileData->tryMatching(newTile);
+		}
+		return {tileData->tileID, matchType};
+	}
+
+	auto size() const { return tiles.size(); }
+
+	auto begin() const { return tiles.begin(); }
+	auto end() const { return tiles.end(); }
+};
+
+static UniqueTiles dedupTiles(Png const &png, DefaultInitVec<AttrmapEntry> &attrmap,
+                              std::vector<Palette> const &palettes,
+                              DefaultInitVec<size_t> const &mappings) {
+	// Iterate throughout the image, generating tile data as we go
+	// (We don't need the full tile data to be able to dedup tiles, but we don't lose anything
+	// by caching the full tile data anyway, so we might as well.)
+	UniqueTiles tiles;
+
+	auto iter = attrmap.begin();
+	for (auto tile : png.visitAsTiles(options.columnMajor)) {
```

Comment:
I've seen at least three parallel iterations over `attrmap` and `png.visitAsTiles`. This isn't a priority, but a custom `zip` iterator might be worth writing (we can't use `range-v3` yet).